modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/snapshot/programs/GlobalSnapshotPulling.scala

```diff
@@ -0,0 +1,86 @@
+package org.tessellation.dag.l1.domain.snapshot.programs
+
+import cats.Applicative
+import cats.data.NonEmptyList
+import cats.effect.Async
+import cats.effect.std.Random
+import cats.syntax.applicativeError._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import org.tessellation.dag.l1.domain.snapshot.storage.LastGlobalSnapshotOrdinalStorage
+import org.tessellation.dag.l1.http.p2p.clients.L0GlobalSnapshotClient
+import org.tessellation.dag.snapshot.{GlobalSnapshot, SnapshotOrdinal}
+import org.tessellation.ext.cats.syntax.next._
+import org.tessellation.sdk.domain.cluster.storage.L0ClusterStorage
+import org.tessellation.security.signature.Signed
+
+import eu.timepit.refined.types.numeric.NonNegLong
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object GlobalSnapshotPulling {
+
+  def make[F[_]: Async: Random](
+    l0GlobalSnapshotClient: L0GlobalSnapshotClient[F],
+    l0ClusterStorage: L0ClusterStorage[F],
+    lastOrdinalStorage: LastGlobalSnapshotOrdinalStorage[F]
+  ): GlobalSnapshotPulling[F] =
+    new GlobalSnapshotPulling[F](l0GlobalSnapshotClient, l0ClusterStorage, lastOrdinalStorage) {}
+}
+
+sealed abstract class GlobalSnapshotPulling[F[_]: Async: Random] private (
+  l0GlobalSnapshotClient: L0GlobalSnapshotClient[F],
+  l0ClusterStorage: L0ClusterStorage[F],
+  lastOrdinalStorage: LastGlobalSnapshotOrdinalStorage[F]
+) {
+
+  private val logger = Slf4jLogger.getLogger[F]
+
+  def pull: F[List[Signed[GlobalSnapshot]]] = {
+    for {
```

Comment:
I did it on purpose and I think it's the good mix. The linear transformations are flatMapped, whatever is used twice (with exception of next) is "named" using for comprehension.

Reply:
yeah, and that's why it's mixed and hard to read. what about following code? it didn't mix for comprehension with flatMaps, and still you achieve what you want, with "named" things. yet, it's easier to follow I think (at least for me).
I'd also split `pulled` into parts, especially `tailRecM` but it's up to you

```
def pull: F[List[Signed[GlobalSnapshot]]] = {
  lastOrdinalStorage.get.flatMap { lastStoredOrdinal =>

    def l0Peer = l0ClusterPeers.getPeers
      .map(_.toNonEmptyList.toList)
      .flatMap(Random[F].shuffleList)
      .map(_.head)

    def lastOrdinal = l0GlobalSnapshotClient.getLatestOrdinal.run(l0Peer)

    def pulled = lastOrdinal.map(_.getOrElse(lastStoredOrdinal))
      .map(last.next.value.value to _.value.value) // btw, I think you can create list of NonNegLong here
      .map(_.toList.map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o))))
      .flatMap { ordinals => .... }
      .map(NonEmptyList.fromList)
      .flatMap { // btw, I'd use lastOrdinalStorage.update to make sure ordinal didn't change in background
        case Some(snapshots) => lastOrdinalStorage.set(snapshots.last.ordinal).map(_ => snapshots.toList)
        case None => List.empty[Signed[GlobalSnapshot]].pure[F]
      }
      
     pulled.handleErrorWith { e => ... }
  }
}
```

Reply:
- There is no `to` function available for `NonNegLong` so I can't use it.
- I can extract `tailRecM` part to a separate function.
- Currently setting the last snapshot ordinal is only used here (that's what it was intended for exclusively), but to future proof it (if some other part of code starts using it) I will compare during set/update if the last known value is equal to the last pulled (kind of optimistic lock) and only then I will return the list of pulled GlobalSnapshots
- Regarding the rest I don't see how naming the parts of code with defs instead of using for comprehension is really better.